pkg/cli/internal/cmd/preflight/checkers.go

```diff
@@ -0,0 +1,167 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package preflight
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gernest/wow"
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	versionutil "k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kubermatic/kubecarrier/pkg/cli/internal/spinner"
+	"github.com/kubermatic/kubecarrier/pkg/internal/util"
+)
+
+var (
+	scheme = runtime.NewScheme()
+)
+
+func init() {
+	utilruntime.Must(appsv1.AddToScheme(scheme))
+	utilruntime.Must(corev1.AddToScheme(scheme))
+}
+
+const (
+	firstSupportedKubernetesVersion = "v1.16.0"
+
+	certManagerNamespaceName            = "cert-manager"
+	certManagerCAInjectorDeploymentName = "cert-manager-cainjector"
+	certManagerWebhookDeploymentName    = "cert-manager-webhook"
+	certManagerDeploymentName           = "cert-manager"
+)
+
+// checker checks if the state of the system meets KubeCarrier installation requirements
+type checker interface {
+	check() error
+	name() string
+}
+
+func RunChecks(c *rest.Config, s *wow.Wow, startTime time.Time, log logr.Logger) error {
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(c)
+	if err != nil {
+		return fmt.Errorf("cannot create discovery client: %w", err)
+	}
+	kubernetesVersion, err := discoveryClient.ServerVersion()
+	if err != nil {
+		return fmt.Errorf("can not get the kubernetesVersion: %w", err)
+	}
+	cl, err := util.NewClientWatcher(c, scheme, log)
+	if err != nil {
+		return fmt.Errorf("creating Kubernetes client: %w", err)
+	}
+	var errBuffer bytes.Buffer
+	checkers := []checker{
+		&kubernetesVersionChecker{
+			firstSupportedVersion: firstSupportedKubernetesVersion,
+			kubernetesVersion:     kubernetesVersion.String(),
+		},
+		&certManagerChecker{
+			client:                 cl,
+			certManagerNamespace:   certManagerNamespaceName,
+			certManagerDeployments: []string{certManagerDeploymentName, certManagerCAInjectorDeploymentName, certManagerWebhookDeploymentName},
+		},
+	}
+	for _, checker := range checkers {
+		if err := spinner.AttachSpinnerTo(s, startTime, fmt.Sprintf("[preflight check] %s", checker.name()), func() error {
+			return checker.check()
+		}); err != nil {
+			errBuffer.WriteString(fmt.Errorf("[preflight check] %s: %w\n", checker.name(), err).Error())
+		}
+	}
+	if errBuffer.Len() > 0 {
+		return fmt.Errorf(errBuffer.String())
+	}
+	return nil
+}
+
+// kubernetesVersionChecker checks if the Kubernetes version of the cluster meets the requirement to deploy KubeCarrier.
+type kubernetesVersionChecker struct {
```

Comment:
This shall soonish start being insufficient. We're probably going to have supported some minor version after specific patch versions. (e.g. 1.16.4+ , 1.17.2+ (but not 1.17.0 nor 1.17.1) ) )

Reply:
I don't really know what you mean? 1.16.4 <= 1.17.2? For this case, it works

Reply:
Do you mean we need to compare some versions like 1.16.4 and 1.17.2?

Reply:
also there is a test about it https://github.com/kubermatic/kubecarrier/pull/313/files#diff-2152533908d9e0e3e13faa9697127d76R64

Reply:
> I don't really know what you mean? 1.16.4 <= 1.17.2? For this case, it works

but it doesn't exclude 1.17.0 and 1.17.1 as an unsupported version, because for example, they don't include some bugfix we need for proper functioning.

Reply:
Ok, I see, should we cover this case also @thetechnick ?

Reply:
Let's validate for the lowest supported version first, we can maintain a blacklist of "broken" k8s versions, if we have a case for it. Currently there is no need to add additional complexity here.